Add executable fall-through

yamadapc opened the pull request:
Uses `System.Directory.findExecutable`. Adds a `mOnInvalidArgs` callback to the
`complicatedOptions` function on `Options.Applicative.Complicated` the rest of
the logic is centralized on `Main`.

I didn't use `System.Process.Read.findExecutable` because I'd then need
`getMinimalEnvOverride`, which would require `GlobalOpts` I think. I'm not sure,
but I don't understand what'd give at the moment.

Will rebase after discussion.

This closes #1433.

yamadapc commented:
Updated to conform to 2 spaces indentation and not run `stylish-haskell` on the files.

borsboom commented on src/main/Main.hs:
```diff
@@ -138,6 +140,16 @@ main = withInterpreterArgs stackProgName $ \args isInterpreter -> do
+           mExternalExec <- Directory.findExecutable ("stack-" ++ head as)
+           case mExternalExec of
+             Just ex -> do
+               ph <- spawnProcess ex (tail as)
```
It would be better if this replaces the current process instead of spawning a sub-process and just waiting for it to finish.  `Stack.Exec.exec` will do that, but it's another function that needs an EnvOverride (this time via a Config).  It's possible to create an EnvOverride using `System.Process.Read.mkEnvOverride`, and `exec` could be modified to take an EnvOverride as an argument instead of needing a Config.

yamadapc commented:
It seems like I'd still need a config to know the platform... I made some changes, but couldn't get this to work.

borsboom commented:
Using Distrubition.System.buildPlatform will suffice for this case.

yamadapc commented:
Cool, thanks for the tip; I should have asked sooner. Updated and rebased against `origin/master`. I made it so that the callback I added to `Complicated.hs` was run on every parser failure; not just arguments. This seemed better.

Not sure if this was the `exec` modification you mentioned?

borsboom commented:
Thanks, this is looking just about ready for merge.  I tested it and everything seems to be in order.  Please add an entry to the changelog, and maybe squash it into one commit.

yamadapc commented:
Okie, coolness! Squashed, changelog entry added and rebased against `master`.
